### PR TITLE
planner: move `rule_constant_propagation` to rule pkg. (#55231)

### LINKED ISSUE #55231: Rearrange logical optimizing rule
## Enhancement

This is a special topic campaign about first-time contributions that are friendly for academic open-source credits, with no hard-core database logic inside. Still, it is a great window for stepping deep into the charm of TiDB optimizer. Here is an example, https://github.com/pingcap/tidb/pull/55226, about moving and classifying current logical rules.

currently, tidb optimizing logical rules are put in the planner/core pkg with the prefix `rule_`, inside these files it may rely on logical operator pkg to do the type check and rule asserting. Some of it may have little coupling with `planner/core` pkg itselves.  When we try to move and classify these rules into `planner/rule` pkg, we should think twice about how to solve these back awards dependencies. The healthy pkg reference graph should be like below:

image

some special cases'(complex rule file migrating) scope will become even bigger than the pictures show, you can contact me  @AilinKid and @hawkingrei to talk about it. 

You are appreciated for making a detailed description of what your pull request has done. Of course, first, you should claim and comment on which rule_file you picked behind this issue, then, after you go through the code, you can modify the same comment to add what's your basic idea about it. After @AilinKid and @hawkingrei approve your idea, you can file your pull request. This is to avoid wasting time on the wrong way too much, describe your idea is much cheaper, isn't it? Thank you so much for reading until here.

### PR DESCRIPTION
**Pull request by qingfeng777**
### What problem does this PR solve?

Issue Number: ref #55231

Problem Summary:

### What changed and how does it work?
move rule_constant_propagation from core to rule pkg

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  > - [x] I ran locally, and executed a sql with pagination.
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by hawkingrei** [comment]
@qingfeng777 Please run the ```make bazel_prepare```.

**Comment by qingfeng777** [comment]
@hawkingrei Done

File: pkg/planner/core/rule/util/BUILD.bazel [hunk 1]
```diff
@@ -5,11 +5,13 @@ go_library(
     srcs = [
         "build_key_info_misc.go",
         "misc.go",
+        "rule_constant_propagation.go",
```
Comment (AilinKid) [inline pkg/planner/core/rule/util/BUILD.bazel:8]: you can rename the file as `constant_propagation_misc.go` and then `make bazel_prepare`

File: pkg/planner/core/rule/util/rule_constant_propagation.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
+// Copyright 2023 PingCAP, Inc.
```
Comment (AilinKid) [inline pkg/planner/core/rule/util/rule_constant_propagation.go:1]: change the date as 2024

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

basically from the workflow, you should add your idea first in the comment then file your pull request after we acked.
like

"idea: after migrating the file,  the function validCompareConstantPredicate is not valid anymore, it is refed by logical selection in the core pkg.   I moved it into `constant_propagation.go` as in ruleutil for usage."

anyway, since this one is quite simple, these changes look good to me.

**Comment by AilinKid** [comment]
`validCompareConstantPredicate` is moved to expression pkg, seems we don't need to handle it here now.

**Comment by qingfeng777** [comment]
> `validCompareConstantPredicate` is moved to expression pkg, seems we don't need to handle it here now.

After my PR, you refactor the codebase. I have deleted it now.

**Review by hawkingrei — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]

**Pull request merged**